src/Bridge/Graphql/Resolver/CollectionResolverFactory.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Bridge\Graphql\Resolver;
+
+use ApiPlatform\Core\Api\IdentifiersExtractorInterface;
+use ApiPlatform\Core\DataProvider\CollectionDataProviderInterface;
+use ApiPlatform\Core\DataProvider\PaginatorInterface;
+use ApiPlatform\Core\DataProvider\SubresourceDataProviderInterface;
+use GraphQL\Type\Definition\ResolveInfo;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+/**
+ * Creates a function retrieving a collection to resolve a GraphQL query.
+ *
+ * @author Alan Poulain <contact@alanpoulain.eu>
+ */
+final class CollectionResolverFactory extends AbstractResolverFactory implements CollectionResolverFactoryInterface
+{
+    private $collectionDataProvider;
+    private $normalizer;
+    private $identifiersExtractor;
+    private $requestStack;
+    private $paginationEnabled;
+
+    public function __construct(CollectionDataProviderInterface $collectionDataProvider, SubresourceDataProviderInterface $subresourceDataProvider, NormalizerInterface $normalizer, IdentifiersExtractorInterface $identifiersExtractor, RequestStack $requestStack, bool $paginationEnabled)
+    {
+        parent::__construct($subresourceDataProvider);
+
+        $this->collectionDataProvider = $collectionDataProvider;
+        $this->normalizer = $normalizer;
+        $this->identifiersExtractor = $identifiersExtractor;
+        $this->requestStack = $requestStack;
+        $this->paginationEnabled = $paginationEnabled;
+    }
+
+    public function createCollectionResolver(string $resourceClass, string $operationName, string $rootClass): callable
+    {
+        return function ($root, $args, $context, ResolveInfo $info) use ($resourceClass, $operationName, $rootClass) {
+            $request = $this->requestStack->getCurrentRequest();
+            if (null !== $request) {
+                $request->attributes->set('_graphql_collections_args', [$resourceClass => $args] + $request->attributes->get('_graphql_collections_args', []));
+            }
+
+            $rootProperty = $info->fieldName;
+            if (isset($root[$rootProperty])) {
+                $rootIdentifiers = $this->identifiersExtractor->getIdentifiersFromResourceClass($rootClass);
+                $subresource = $this->getSubresource($rootClass, $root, $rootIdentifiers, $rootProperty, $resourceClass, $operationName, true);
+                $collection = $subresource ?? [];
+            } else {
+                $collection = $this->collectionDataProvider->getCollection($resourceClass, $operationName);
+            }
+
+            if ($this->paginationEnabled) {
```

Comment:
I would move all the following block in a new method to improve readability and decrease the cyclomatic complexity.